modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/IO.kt

```diff
@@ -220,6 +235,25 @@ sealed class IO<out A> : IOOf<A> {
 
     override fun unsafeRunTimedTotal(limit: Duration): Option<A> = throw AssertionError("Unreachable")
   }
+
+  internal data class MapFilter<E, out A>(val source: IOOf<E>, val g: (E) -> Option<A>, val index: Int) : IO<A>(), (E) -> IO<A> {
+    override fun invoke(value: E): IO<A> = g(value)
+      .fold(
+        { raiseError(IllegalArgumentException("IO execution should yield a valid result")) },
+        { just(it) }
+      )
+
+    override fun <B> mapFilter(f: (A) -> Option<B>): IO<B> {
+      val h: (Option<A>) -> Option<B> = { it.fold({ None }, { a: A -> f(a) }) }
+
+      return when {
+        index != maxStackDepthSize -> MapFilter(source, g.andThen(h), index + 1)
+        else -> MapFilter(this, f, 0)
```

Comment:
Only define `h` if it's going to be used, else it's an unnecessary allocation

Reply:
Removed after refactor